pkg/upgrade/upgrade.go

```diff
@@ -13,18 +13,47 @@
 
 package upgrade
 
-import "go.etcd.io/etcd/clientv3"
+import (
+	"context"
+	"fmt"
+
+	"github.com/pingcap/tidb-tools/pkg/dbutil"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/conn"
+	tcontext "github.com/pingcap/dm/pkg/context"
+	"github.com/pingcap/dm/pkg/cputil"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/utils"
+)
 
 var (
 	// upgrades records all functions used to upgrade from one version to the later version.
-	upgrades = []func(cli *clientv3.Client) error{
+	upgrades = []func(cli *clientv3.Client, uctx Context) error{
 		upgradeToVer1,
+		upgradeToVer2,
 	}
 )
 
+// Context is used to pass something to TryUpgrade
+// NOTE that zero value of Context is nil, be aware of nil-dereference
+type Context struct {
+	context.Context
+	SubTaskConfigs map[string]map[string]config.SubTaskConfig
+}
+
+// NewUpgradeContext creates a Context, avoid nil Context member
+func NewUpgradeContext() Context {
+	return Context{
+		Context: context.Background(),
```

Comment:
Should we need to `make` a `map` for it?

Reply:
https://github.com/pingcap/dm/pull/988/commits/3ee0d8c0c3075ff5b5c2cef45e38fbf87a61238c